### PR TITLE
chore(launch): Warn when wandb is missing from requirements

### PR DESCRIPTION
**Pull request by TimH98**
Description
-----------

- Fixes [WB-18376](https://wandb.atlassian.net/browse/WB-18376)

Display a warning when the user tries to create or launch a job where `wandb` isn't in the `requirements.txt`, since it's generally required for a run to succeed.

- [x] I updated CHANGELOG.md, or it's not applicable

Testing
-------
Made a code job with no wandb in the `requirements.txt`

[WB-18376]: https://wandb.atlassian.net/browse/WB-18376?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by bcsherma — COMMENTED** [review]
I don't think `wandb` will usually be in the requirements line? The requirements line is often just `pip install -r src/requirements.txt` or something like that. You need to check the actual `requirements.txt` file if it exists

It looks like the warn path  `_project_spec.py` path is being hit by tests (which is a lil weird bc I don't think we have a test case for this) and the `create_job.py` is not covered. Can we add tests for these?

File: wandb/sdk/launch/_project_spec.py [hunk 1]
```diff
@@ -486,6 +486,8 @@ def parse_existing_requirements(self) -> str:
                         _logger.warn(f"Unable to parse requirements.txt: {e}")
                         continue
             requirements_line += "WANDB_ONLY_INCLUDE={} ".format(",".join(include_only))
+            if "wandb" not in requirements_line:
+                wandb.termwarn(f"{LOG_PREFIX}wandb is not present in requirements.txt.")
```
Comment (TimH98) [inline wandb/sdk/launch/_project_spec.py:490]: @bcsherma I think this is good as-is, look at how `requirements_line` is built from lines 470-488. The contents of `requirements.txt` are added to `include_only` which in turn is included in `requirements_line`.

Comment (bcsherma) [reply wandb/sdk/launch/_project_spec.py:490]: try launching this job from your branch https://wandb.ai/bcanfieldsherman/uncategorized/artifacts/job/job-source-uncategorized-main.py/v8

it has a code artifact with a requirements.txt in it that doesn't have `wandb` in it. If you launch it locally with your branch you should see something like:

```
wandb: launch: Launching run into bcanfieldsherman/uncategorized
wandb: launch: Launching job: bcanfieldsherman/uncategorized/job-source-uncategorized-main.py:v8
wandb:   2 of 2 files downloaded.  
wandb:   2 of 2 files downloaded.  
#0 building with "desktop-linux" instance using docker driver

#1 [internal] load build definition from Dockerfile.wandb
#1 transferring dockerfile: 989B done
#1 DONE 0.0s

#2 [internal] load metadata for docker.io/library/python:3.10
#2 ...

#3 [internal] load metadata for docker.io/library/python:3.10-buster
#3 DONE 0.4s

#2 [internal] load metadata for docker.io/library/python:3.10
#2 DONE 0.4s

#4 [internal] load .dockerignore
#4 transferring context: 2B done
#4 DONE 0.0s

#5 [build 1/4] FROM docker.io/library/python:3.10@sha256:817c0d8684087acb6d88f0f0951f9a541aa3e762302aa5e8f439d5d12edd48ad
#5 DONE 0.0s

#6 [base 1/7] FROM docker.io/library/python:3.10-buster@sha256:31a8498af27dba1f431f2df60189ca79ff5f440cfa7b6dea2da6ab15c74abcd4
#6 DONE 0.0s

#7 [build 2/4] RUN python -m venv /env
#7 CACHED

#8 [internal] settings cache mount permissions
#8 CACHED

#9 [internal] load build context
#9 transferring context: 7.93kB done
#9 DONE 0.0s

#10 [build 3/4] COPY src/requirements.txt ./
#10 DONE 0.0s

#11 [build 4/4] RUN --mount=type=cache,mode=0777,target=/root/.cache/pip pip install -r requirements.txt
#11 0.478 Collecting numpy
#11 0.509   Using cached numpy-1.26.4-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.whl (14.2 MB)
#11 0.556 Installing collected packages: numpy
#11 1.492 Successfully installed numpy-1.26.4
#11 1.495 WARNING: There was an error checking the latest version of pip.
#11 DONE 1.6s

#6 [base 1/7] FROM docker.io/library/python:3.10-buster@sha256:31a8498af27dba1f431f2df60189ca79ff5f440cfa7b6dea2da6ab15c74abcd4
#6 CACHED

#12 [base 2/7] COPY --from=build /env /env
#12 DONE 0.2s

#13 [base 3/7] RUN useradd     --create-home     --no-log-init     --shell /bin/bash     --gid 0     --uid 1000     ben || echo ""
#13 DONE 0.1s

#14 [base 4/7] WORKDIR /home/ben
#14 DONE 0.0s

#15 [base 5/7] RUN chown -R 1000 /home/ben
#15 DONE 0.1s

#16 [base 6/7] RUN mkdir -p /home/ben/.cache && chown -R 1000 /home/ben/.cache
#16 DONE 0.2s

#17 [base 7/7] COPY --chown=1000 src/ /home/ben
#17 DONE 0.0s

#18 exporting to image
#18 exporting layers
#18 exporting layers 0.2s done
#18 writing image sha256:fe5a07e08ddfe44171422d4142bc646d8a326fe55173fe613711fb22817315e9 done
#18 naming to docker.io/library/bcanfieldsherman__uncategorized__job-source-uncategorized-main.py:56ae2ca3
#18 naming to docker.io/library/bcanfieldsherman__uncategorized__job-source-uncategorized-main.py:56ae2ca3 done
#18 DONE 0.2s

View build details: docker-desktop://dashboard/build/desktop-linux/desktop-linux/lm8h37jl42f8hsx2kwdw534m6
#0 building with "desktop-linux" instance using docker driver

#1 [internal] load build definition from Dockerfile.wandb
#1 transferring dockerfile: 989B done
#1 DONE 0.0s

#2 [internal] load metadata for docker.io/library/python:3.10
#2 ...

#3 [internal] load metadata for docker.io/library/python:3.10-buster
#3 DONE 0.4s

#2 [internal] load metadata for docker.io/library/python:3.10
#2 DONE 0.4s

#4 [internal] load .dockerignore
#4 transferring context: 2B done
#4 DONE 0.0s

#5 [build 1/4] FROM docker.io/library/python:3.10@sha256:817c0d8684087acb6d88f0f0951f9a541aa3e762302aa5e8f439d5d12edd48ad
#5 DONE 0.0s

#6 [base 1/7] FROM docker.io/library/python:3.10-buster@sha256:31a8498af27dba1f431f2df60189ca79ff5f440cfa7b6dea2da6ab15c74abcd4
#6 DONE 0.0s

#7 [build 2/4] RUN python -m venv /env
#7 CACHED

#8 [internal] settings cache mount permissions
#8 CACHED

#9 [internal] load build context
#9 transferring context: 7.93kB done
#9 DONE 0.0s

#10 [build 3/4] COPY src/requirements.txt ./
#10 DONE 0.0s

#11 [build 4/4] RUN --mount=type=cache,mode=0777,target=/root/.cache/pip pip install -r requirements.txt
#11 0.478 Collecting numpy
#11 0.509   Using cached numpy-1.26.4-cp310-cp310-manylinux_2_17_aarch64.manylinux2014_aarch64.whl (14.2 MB)
#11 0.556 Installing collected packages: numpy
#11 1.492 Successfully installed numpy-1.26.4
#11 1.495 WARNING: There was an error checking the latest version of pip.
#11 DONE 1.6s

#6 [base 1/7] FROM docker.io/library/python:3.10-buster@sha256:31a8498af27dba1f431f2df60189ca79ff5f440cfa7b6dea2da6ab15c74abcd4
#6 CACHED

#12 [base 2/7] COPY --from=build /env /env
#12 DONE 0.2s

#13 [base 3/7] RUN useradd     --create-home     --no-log-init     --shell /bin/bash     --gid 0     --uid 1000     ben || echo ""
#13 DONE 0.1s

#14 [base 4/7] WORKDIR /home/ben
#14 DONE 0.0s

#15 [base 5/7] RUN chown -R 1000 /home/ben
#15 DONE 0.1s

#16 [base 6/7] RUN mkdir -p /home/ben/.cache && chown -R 1000 /home/ben/.cache
#16 DONE 0.2s

#17 [base 7/7] COPY --chown=1000 src/ /home/ben
#17 DONE 0.0s

#18 exporting to image
#18 exporting layers
#18 exporting layers 0.2s done
#18 writing image sha256:fe5a07e08ddfe44171422d4142bc646d8a326fe55173fe613711fb22817315e9 done
#18 naming to docker.io/library/bcanfieldsherman__uncategorized__job-source-uncategorized-main.py:56ae2ca3
#18 naming to docker.io/library/bcanfieldsherman__uncategorized__job-source-uncategorized-main.py:56ae2ca3 done
#18 DONE 0.2s

View build details: docker-desktop://dashboard/build/desktop-linux/desktop-linux/lm8h37jl42f8hsx2kwdw534m6

wandb: launch: Launching run in docker with command: docker run --rm -e WANDB_BASE_URL=https://api.wandb.ai -e WANDB_API_KEY -e WANDB_PROJECT=uncategorized -e WANDB_ENTITY=bcanfieldsherman -e WANDB_LAUNCH=True -e WANDB_RUN_ID=cg72qbu5 -e WANDB_CONFIG='{}' -e WANDB_LAUNCH_FILE_OVERRIDES='{}' -e WANDB_ARTIFACTS='{"_wandb_job": "bcanfieldsherman/uncategorized/job-source-uncategorized-main.py:v8"}' bcanfieldsherman__uncategorized__job-source-uncategorized-main.py:56ae2ca3
Traceback (most recent call last):
  File "/home/ben/main.py", line 1, in <module>
    import wandb
ModuleNotFoundError: No module named 'wandb'
wandb: ERROR Launched run exited with non-zero status
```

and as you can see there is no warning and the requirements line does not have the text `wandb` in it

Comment (bcsherma) [reply wandb/sdk/launch/_project_spec.py:490]: The `wandb job create` or `wandb launch -u` path works correctly but if the job is made from a run and there is a user provided `requirements.txt` in the build context I don't think this is correct

Comment (TimH98) [reply wandb/sdk/launch/_project_spec.py:490]: Oh good catch thanks, looks like I missed a case in `build.py`

**Pull request merged**